src/huggingface_hub/hf_api.py

```diff
@@ -2245,6 +2264,674 @@ def get_full_repo_name(
+            conflicting_files=discussion_details["filesWithConflicts"]
+            if is_pull_request
+            else None,
+            target_branch=discussion_details["changes"]["base"]
+            if is_pull_request
+            else None,
+            merge_commit_oid=discussion_details["changes"].get("mergeCommitId", None)
+            if is_pull_request
+            else None,
+            diff=discussion_details.get("diff"),
+        )
+
+    def create_discussion(
+        self,
+        repo_id: str,
+        title: str,
+        *,
+        token: str,
+        description: Optional[str] = None,
+        repo_type: Optional[str] = None,
+        pull_request: bool = False,
+    ) -> DiscussionWithDetails:
+        """Creates a discussion or pull request.
+
+        Pull Requests created programmatically will be in `"draft"` status.
+
+        Args:
+            repo_id (`str`):
+                A namespace (user or an organization) and a repo name separated
+                by a `/`.
+            title (`str`):
+                The title of the discussion. It can be up to 200 characters long,
+                and must be at least 3 characters long. Leading and trailing whitespaces
+                will be stripped.
+            token (`str`):
+                An authentication token (See https://huggingface.co/settings/token)
+            description (`str`, *optional*):
+                An optional description for the pull request.
+                Defaults to `"Discussion opened with the huggingface_hub Python library"`
+            pull_request (`bool`, *optional*):
+                Whether to create a pull request or discussion. If `True`, creates a pull request.
+                If `False`, creates a discussion. Defaults to `False`.
+            repo_type (`str`, *optional*):
+                Set to `"dataset"` or `"space"` if uploading to a dataset or
+                space, `None` or `"model"` if uploading to a model. Default is
+                `None`.
+
+        Returns: [`DiscussionWithDetails`]
+
+        <Tip>
+
+        Raises the following errors:
+
+            - [`HTTPError`](https://2.python-requests.org/en/master/api/#requests.HTTPError)
+              if the HuggingFace API returned an error
+            - [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError)
+              if some parameter value is invalid
+            - [`~huggingface_hub.utils.RepositoryNotFoundError`]
+              If the repository to download from cannot be found. This may be because it doesn't exist,
+              or because it is set to `private` and you do not have access.
+
+        </Tip>"""
+        if repo_type not in REPO_TYPES:
+            raise ValueError(f"Invalid repo type, must be one of {REPO_TYPES}")
+        if repo_type is None:
+            repo_type = REPO_TYPE_MODEL
+        full_repo_id = f"{repo_type}s/{repo_id}"
+        if token is None:
+            token = HfFolder.get_token()
```

Comment:
shouldn't we use `self._validate_or_retrieve_token` instead?

Reply:
Oh yes definitely!